airbyte-integrations/connectors/source-shopify/source_shopify/transform.py

```diff
@@ -0,0 +1,116 @@
+    Correct types placed in schemes
+    Transformer iterates over records, compare values type with schema type and transform if it's needed
+
+    Methods
+    -------
+    _transform_array(self, array: List[Any], item_properties: Mapping[str, Any])
+        Some fields type is array. Items inside array contain price fields, which should be transformed
+        This method iterate over items in array, compare schema types and convert if necessary
+    """
+
+    def __init__(self, schema: Mapping[str, Any], **kwargs):
+        super().__init__(**kwargs)
+        self._schema = schema
+
+    @staticmethod
+    def _get_json_types(value_type) -> List[str]:
+        json_types = {
+            str: ["string"],
+            int: ["integer", "number"],
+            float: ["number"],
+            dict: ["object"],
+            list: ["array"],
+            bool: ["boolean"],
+            type(None): [
+                "null",
+            ],
+        }
+        return json_types.get(value_type)
+
+    @staticmethod
+    def _types_from_schema(properties: Mapping[str, Any]) -> str:
+        schema_types = properties.get("type", [])
+        if not isinstance(schema_types, list):
+            schema_types = [
+                schema_types,
+            ]
+        return schema_types
+
+    @staticmethod
+    def _first_non_null_type(schema_types: List[str]) -> str:
+        not_null_types = schema_types.copy()
+        if "null" in not_null_types:
+            not_null_types.remove("null")
+        return not_null_types[0]
+
+    @staticmethod
+    def _transform_number(value: Any):
+        return Decimal(value)
+
+    def _transform_array(self, array: List[Any], item_properties: Mapping[str, Any]):
+        # iterate over items in array, compare schema types and convert if necessary.
+        for index, record in enumerate(array):
+            array[index] = self.transform(record, item_properties)
+        return array
+
+    def _transform_object(self, record: MutableMapping[str, Any], properties: Mapping[str, Any]):
+        # compare schema types and convert if necessary.
+        for object_property, value in record.items():
+            if value is None:
+                continue
+            if object_property in properties:
+                object_properties = properties.get(object_property) or {}
+                record[object_property] = self.transform(value, object_properties)
+        return record
+
+    def transform(self, record: MutableMapping[str, Any], schema: Mapping[str, Any] = None) -> Iterable[MutableMapping]:
+        schema = schema if schema is not None else self._schema
+        schema_types = self._types_from_schema(schema)
+        if schema_types:
+            value_json_types = self._get_json_types(type(record))
+            schema_type = self._first_non_null_type(schema_types)
+            if not any(value_json_type in schema_types for value_json_type in value_json_types):
```

Comment:
can you add source code comments for what is happening? It's really unclear what check this branch is performing. This is particularly confusing because `record`'s type signature is a `MutableMapping` which makes me think it will always be a `dict`?

Reply:
looking more at the code it seems like `record` can actually be of any type? maybe it should be called `field` instead?